From 23719c98c248ed9d9507b60eab2672fb6f754b3b Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: Tue, 03 Feb 2026 23:58:20 +0100
Subject: [PATCH 5/35] contrib/features: age: Add password-store integration



---
diff --git a/src/contrib/features/age.scm b/src/contrib/features/age.scm
index 5a234a5..2de667b 100644
--- a/src/contrib/features/age.scm
+++ b/src/contrib/features/age.scm
@@ -1,6 +1,6 @@
 ;;; rde --- Reproducible development environment.
 ;;;
-;;; Copyright © 2023 Nicolas Graves <ngraves@ngraves.fr>
+;;; Copyright © 2023, 2024, 2025 Nicolas Graves <ngraves@ngraves.fr>
 ;;;
 ;;; This file is part of rde.
 ;;;
@@ -19,10 +19,12 @@
 
 (define-module (contrib features age)
   #:use-module (rde features)
+  #:use-module (rde features emacs)
   #:use-module (gnu packages golang-crypto)
   #:use-module (gnu services)
   #:use-module (gnu services base)
   #:use-module (gnu home services)
+  #:use-module (gnu packages emacs-xyz)
   #:use-module (guix gexp)
 
   #:export (feature-age))
@@ -30,24 +32,67 @@
 (define* (feature-age
           #:key
           (age age)
-          (age-ssh-key "~/.ssh/id_ed25519"))
+          (age-ssh-key "~/.ssh/id_ed25519")
+          (emacs-age emacs-age)
+          (emacs-passage emacs-passage))
   "This feature sets up age for encryption tasks. It uses the SSH key
 given in the location age-ssh-key to encrypt and decrypt files in the
-password-store. The feature is in contrib because it currently has limited
-functionality compared to GnuPG, which is the recommended privacy tool for
-RDE."
+password-store.
+
+The feature is in contrib because it currently has limited functionality
+compared to GnuPG.  In particular, Age is currently not able to replace GPG
+for its intended purpose (maintaining a web of trust with folks that you
+communicate with), but provides a lighter weight alternative for almost all
+other uses of Emacs encryption."
   (ensure-pred file-like? age)
   (ensure-pred string? age-ssh-key)
+  (ensure-pred file-like? emacs-age)
+  (ensure-pred file-like? emacs-passage)
+
+  (define emacs-f-name 'passage)
+  (define f-name (symbol-append 'emacs- emacs-f-name))
 
   (define (get-home-services config)
+    (append
+     (list (simple-service
+            'age-add-age-package
+            home-profile-service-type
+            (list (get-value 'age config))))
+     (if (get-value 'pass config)
+         (list
+          (simple-service
+           'passage-environment-variables
+           home-environment-variables-service-type
+           `(("PASSAGE_AGE"
+              . ,(file-append (get-value 'age config) "/bin/age"))
+             ("PASSAGE_DIR"
+              . ,(get-value 'password-store-directory config))
+             ("PASSAGE_IDENTITIES_FILE" . ,(get-value 'age-ssh-key config))
+             ("PASSAGE_RECIPIENTS_FILE"
+              . ,(string-append (get-value 'age-ssh-key config) ".pub")))))
+         '())
+     (get-home-emacs-services config)))
+
+  (define (get-home-emacs-services config)
     (list
-     (simple-service
-      'age-add-age-package
-      home-profile-service-type
-      (list age))))
+     (rde-elisp-configuration-service
+      'emacs-age-authinfo
+      config
+      `((eval-when-compile
+         (require 'age))
+        (autoload 'age-file-enable "age")
+        (age-file-enable)
+        (require 'passage))
+      #:summary "age authentication information emacs interface"
+      #:authors '("Nicolas Graves <ngraves@ngraves.fr>")
+      #:keywords '(convenience)
+      #:elisp-packages (list (get-value 'emacs-age config)
+                             (get-value 'emacs-passage config)))))
 
   (feature
    (name 'age)
    (values `((age . ,age)
-             (age-ssh-key . ,age-ssh-key)))
+             (age-ssh-key . ,age-ssh-key)
+             (emacs-age . ,emacs-age)
+             (emacs-passage . ,emacs-passage)))
    (home-services-getter get-home-services)))
diff --git a/src/rde/features/password-utils.scm b/src/rde/features/password-utils.scm
index b082a31..0da993c 100644
--- a/src/rde/features/password-utils.scm
+++ b/src/rde/features/password-utils.scm
@@ -32,6 +32,7 @@
   #:use-module (gnu services)
 
   #:use-module (guix gexp)
+  #:use-module (guix packages)
 
   #:export (feature-password-store))
 
@@ -62,7 +63,8 @@
 
   (define (password-store-home-services config)
     "Returns home services related to password-store."
-    (require-value 'gpg-primary-key config)
+    (when (string= (package-name password-store) "password-store")
+      (require-value 'gpg-primary-key config))
     (require-value 'home-directory config)
     (list (service home-password-store-service-type
                    (home-password-store-configuration
@@ -208,6 +210,7 @@ Keybinding for `rde-consult-pass' and embark actions for it."
    (name 'password-store)
    (values `((pass . #t)
              (password-store . ,password-store)
+             (password-store-directory . ,password-store-directory)
              ,@(if default-pass-prompt?
                    `((default-pass-prompt-fn . ,emacs-pass-prompt))
                    '())))

--
1.9.1
