From 7f4169398b9b2cce736137e070a45586c548e0ef Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: mer., 04 févr. 2026 15:08:41 +0100
Subject: [PATCH 16/35] feature: python: Add nrepl support



---
diff --git a/files/emacs/arei-python.el b/files/emacs/arei-python.el
index e60b484..36fd74b 100644
--- a/files/emacs/arei-python.el
+++ b/files/emacs/arei-python.el
@@ -48,8 +48,6 @@ Use PARAMS to determine the name of the connection buffer."
           'face
           '((t (:inherit font-lock-comment-face)))))))))
 
-(advice-add 'arei--request-user-eval :after #'message)
-
 (defun arei-python-send-statement ()
   "Send the statement at point to inferior Python process.
 The statement is delimited by `python-nav-beginning-of-statement' and
@@ -110,7 +108,6 @@ The statement is delimited by `python-nav-beginning-of-block' and
                     :after #'arei-python-send-path))
     (arei-mode -1)))
 
-(advice-add 'arei-client--connect :after #'arei-python-send-path)
 
 (provide 'arei-python)
 ;;; arei-python.el ends here
diff --git a/src/rde/features/python.scm b/src/rde/features/python.scm
index 0b3836d..a15e358 100644
--- a/src/rde/features/python.scm
+++ b/src/rde/features/python.scm
@@ -27,6 +27,7 @@
   #:use-module (gnu packages python)
   #:use-module (gnu packages python-xyz)
   #:use-module (gnu packages emacs-xyz)
+  #:use-module (rde packages emacs-xyz)
   #:use-module (guix gexp)
   #:export (feature-python))
 
@@ -76,6 +77,8 @@ if readline.get_current_history_length() == 0:
           (python-debugpy python-debugpy)
           (python-lsp-server python-lsp-server)
           (python-lsp-server-plugins (list python-lsp-black))
+          (emacs-arei-python emacs-arei-python)
+          (nrepl? #t)
           (lsp? #t)
           (dap? #t)
           (max-line-length 88))
@@ -93,6 +96,8 @@ integration for dap."
   (ensure-pred file-like? python-debugpy)
   (ensure-pred file-like? python-lsp-server)
   (ensure-pred list-of-file-likes? python-lsp-server-plugins)
+  (ensure-pred file-like? emacs-arei-python)
+  (ensure-pred boolean? nrepl?)
   (ensure-pred boolean? lsp?)
   (ensure-pred boolean? dap?)
   (ensure-pred integer? max-line-length)
@@ -182,6 +187,10 @@ max-line-length = ~a~%"
              (add-hook 'python-mode-hook 'rde-initialize-guix-python-environment)
              (add-hook 'python-ts-mode-hook 'rde-initialize-guix-python-environment)
 
+             ,@(if nrepl?
+                   `((eval-when-compile (require 'arei-python)))
+                   '())
+
              ,@(if (get-value 'emacs-org config #f)
                    `((with-eval-after-load 'org
                        (add-to-list 'org-structure-template-alist
@@ -191,7 +200,8 @@ max-line-length = ~a~%"
                      (with-eval-after-load 'ob-python
                        (setq org-babel-python-command
                              ,(file-append python "/bin/python"))))
-                   '()))))
+                   '()))
+           #:elisp-packages (if nrepl? (list emacs-arei-python) '())))
          '())))
 
   (feature

--
1.9.1
