From d66b796b3811e4253102bcd08b566f24c36d6180 Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: Wed, 04 Feb 2026 15:08:41 +0100
Subject: [PATCH 19/35] features: emacs-git: Add native support for emacs-git-email



---
diff --git a/src/rde/features/emacs-xyz.scm b/src/rde/features/emacs-xyz.scm
index 6abd7a9..081782f 100644
--- a/src/rde/features/emacs-xyz.scm
+++ b/src/rde/features/emacs-xyz.scm
@@ -3392,6 +3392,7 @@ language for GNU Emacs."
           (emacs-magit-todos emacs-magit-todos)
           (emacs-git-timemachine emacs-git-timemachine)
           (emacs-git-link emacs-git-link)
+          (emacs-git-email emacs-git-email-sans-mu4e)
           (emacs-git-gutter-fringe emacs-git-gutter-fringe)
           (emacs-git-gutter-transient emacs-git-gutter-transient))
   "Configure git-related utilities for GNU Emacs, including magit,
@@ -3402,6 +3403,7 @@ git-link, git-timemachine."
   (ensure-pred file-like? emacs-magit-todos)
   (ensure-pred file-like? emacs-git-timemachine)
   (ensure-pred file-like? emacs-git-link)
+  (ensure-pred file-like? emacs-git-email)
   (ensure-pred file-like? emacs-git-gutter-fringe)
   (ensure-pred file-like? emacs-git-gutter-transient)
 
@@ -3468,6 +3470,10 @@ git-link, git-timemachine."
          (defvar rde-projects-directory ,(or project-directory 'nil)
            "Directory where project repositories are stored.")
 
+         (eval-when-compile
+          (require 'git-email-magit))
+         (git-email-magit-setup)
+
          (autoload 'git-link--parse-remote "git-link")
          (defun rde-get-local-repo-path-from-url (url)
            "Get directory from repository url and suggest it to
@@ -3520,6 +3526,7 @@ Almost all other operations are covered by magit."
       #:keywords '(convenience faces)
       #:elisp-packages (list emacs-magit emacs-magit-todos
                              emacs-git-link emacs-git-timemachine
+                             emacs-git-email
                              emacs-git-gutter-fringe
                              emacs-git-gutter-transient))))
 

--
1.9.1
