From: Nicolas Graves <ngraves@ngraves.fr>
Subject: [PATCH 21/27] features: emacs-git: Support git submodules by default



---
diff --git a/src/rde/features/emacs-xyz.scm b/src/rde/features/emacs-xyz.scm
index 0af4f5d..5f5132d 100644
--- a/src/rde/features/emacs-xyz.scm
+++ b/src/rde/features/emacs-xyz.scm
@@ -3587,8 +3587,9 @@ git-link, git-timemachine."
         (with-eval-after-load
          'magit
          (magit-add-section-hook 'magit-status-sections-hook
-                                 'magit-insert-local-branches
-                                 'magit-insert-stashes)
+                                 'magit-insert-modules
+                                 'magit-insert-stashes
+                                 'append)
          (defvar rde-projects-directory ,(or project-directory 'nil)
            "Directory where project repositories are stored.")
 

--
1.9.2
