From 325d1cea7dd637eff3e6e54fd8dc197c87573d77 Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: Wed, 25 Mar 2026 13:13:22 +0100
Subject: [PATCH 22/28] features: emacs-git: Support git submodules by default



---
diff --git a/src/rde/features/emacs-xyz.scm b/src/rde/features/emacs-xyz.scm
index 49c268c..5698644 100644
--- a/src/rde/features/emacs-xyz.scm
+++ b/src/rde/features/emacs-xyz.scm
@@ -3581,8 +3581,9 @@ git-link, git-timemachine."
         (with-eval-after-load
          'magit
          (magit-add-section-hook 'magit-status-sections-hook
-                                 'magit-insert-local-branches
-                                 'magit-insert-stashes)
+                                 'magit-insert-modules
+                                 'magit-insert-stashes
+                                 'append)
          (defvar rde-projects-directory ,(or project-directory 'nil)
            "Directory where project repositories are stored.")
 

--
1.9.2
