From: Nicolas Graves <ngraves@ngraves.fr>
Subject: [PATCH 23/28] system: install: Add preamble



---
diff --git a/src/rde/system/install.scm b/src/rde/system/install.scm
index 22280d9..5b60517 100644
--- a/src/rde/system/install.scm
+++ b/src/rde/system/install.scm
@@ -1,3 +1,10 @@
+;;; rde --- Reproducible development environment.
+;;;
+;;; SPDX-FileCopyrightText: 2021 Andrew Tropin <andrew@trop.in>
+;;; SPDX-FileCopyrightText: 2024 Kirill Yermak <kirill@kimimii.org>
+;;;
+;;; SPDX-License-Identifier: GPL-3.0-or-later
+
 (define-module (rde system install)
   #:use-module (gnu)
   #:use-module (gnu packages linux)

--
1.9.2
