From 56a3e138c93c1c4c83c2e3425fd3e6e6e701cda0 Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: mer., 04 févr. 2026 15:08:41 +0100
Subject: [PATCH 26/35] features: emacs-git: Support git submodules by default



---
diff --git a/src/rde/features/emacs-xyz.scm b/src/rde/features/emacs-xyz.scm
index cbf208f..2f21bc3 100644
--- a/src/rde/features/emacs-xyz.scm
+++ b/src/rde/features/emacs-xyz.scm
@@ -3465,8 +3465,9 @@ git-link, git-timemachine."
         (with-eval-after-load
          'magit
          (magit-add-section-hook 'magit-status-sections-hook
-                                 'magit-insert-local-branches
-                                 'magit-insert-stashes)
+                                 'magit-insert-modules
+                                 'magit-insert-stashes
+                                 'append)
          (defvar rde-projects-directory ,(or project-directory 'nil)
            "Directory where project repositories are stored.")
 

--
1.9.1
