From 7d3602e5c9117a2392ee5967217c24a258638cb5 Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: Wed, 04 Feb 2026 15:08:41 +0100
Subject: [PATCH 30/35] packages: Remove unused modules



---
diff --git a/src/rde/packages.scm b/src/rde/packages.scm
index b91f478..f4bc091 100644
--- a/src/rde/packages.scm
+++ b/src/rde/packages.scm
@@ -35,14 +35,11 @@
 
   #:use-module (srfi srfi-1)
 
-  #:use-module (guix diagnostics)
-  #:use-module (guix i18n)
   #:use-module (guix packages)
   #:use-module (guix gexp)
   #:use-module (guix utils)
   #:use-module (guix download)
   #:use-module (guix git-download)
-  #:use-module (guix build-system meson)
   #:use-module (guix build-system gnu)
   #:use-module (guix build-system guile)
   #:use-module ((guix licenses) #:prefix license:)

--
1.9.1
