From 27bcc60a0725f3fbd5f7b582f51c681c0c9e1f1d Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: mer., 04 févr. 2026 15:13:31 +0100
Subject: [PATCH 33/35] feature: emacs: Migrate to emacs-xdg-launcher.



---
diff --git a/src/rde/features/emacs.scm b/src/rde/features/emacs.scm
index 21f8463..c8dced6 100644
--- a/src/rde/features/emacs.scm
+++ b/src/rde/features/emacs.scm
@@ -599,8 +599,8 @@ It can contain settings not yet moved to separate features."
 
   (define (emacs-application-launcher config)
     ((emacs-minibuffer-program config)
-     "application-launcher" "Application Launcher"
-     'app-launcher-run-app #:height standalone-minibuffer-height))
+     "xdg-application-launcher" "XDG Application Launcher"
+     'xdg-launcher-run-app #:height standalone-minibuffer-height))
 
   (define (emacs-home-services config)
     "Returns home services related to GNU Emacs."
@@ -627,7 +627,7 @@ It can contain settings not yet moved to separate features."
        (elisp-packages
         (append
          additional-elisp-packages
-         (if default-application-launcher? (list emacs-app-launcher) '())))
+         (if default-application-launcher? (list emacs-xdg-launcher) '())))
        (emacs-servers (if emacs-server-mode? '(server) '()))
        (xdg-flavor? #t)
        (early-init-el

--
1.9.1
